common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -655,6 +747,194 @@ public void registerExecutor(String appId, ExecutorShuffleInfo executorInfo) {
+      return keyJsonString.getBytes(StandardCharsets.UTF_8);
+    } catch (Exception exception) {
+      logger.error("Exception while generating the DB key {}", key);
+      return null;
+    }
+  }
+
+  private byte[] getDbAppAttemptShufflePartitionKey(
+      AppAttemptShuffleMergeId appAttemptShuffleMergeId) {
+    return getDbKey(appAttemptShuffleMergeId, APP_ATTEMPT_SHUFFLE_FINALIZE_STATUS_KEY_PREFIX);
+  }
+
+  private byte[] getDbAppAttemptPathsKey(AppAttemptId appAttemptId){
+    return getDbKey(appAttemptId, APP_ATTEMPT_PATH_KEY_PREFIX);
+  }
+
+  @VisibleForTesting
+  void reloadAndCleanUpAppShuffleInfo(DB db) {
+    logger.info("Reload applications merged shuffle information from DB");
+    List<byte[]> dbKeysToBeRemoved = new ArrayList<>();
+    reloadActiveAppAttemptsPathInfo(db, dbKeysToBeRemoved);
+    reloadFinalizedAppAttemptsShuffleMergeInfo(db, dbKeysToBeRemoved);
+    // Clean up invalid data stored in DB
+    mergedShuffleCleaner.execute(() ->
+        dbKeysToBeRemoved.forEach(
+            (key) -> {
+              try {
+                db.delete(key);
+              } catch (Exception e) {
+                logger.error("Error deleting data in DB", e);
+              }
+            }
+        )
+    );
+  }
+
+  private void reloadActiveAppAttemptsPathInfo(DB db, List<byte[]> dbKeysToBeRemoved) {
+    if (db != null) {
+      DBIterator itr = db.iterator();
+      itr.seek(APP_ATTEMPT_PATH_KEY_PREFIX.getBytes(StandardCharsets.UTF_8));
+      while (itr.hasNext()) {
+        Map.Entry<byte[], byte[]> e = itr.next();
+        String key = new String(e.getKey(), StandardCharsets.UTF_8);
+        if (!key.startsWith(APP_ATTEMPT_PATH_KEY_PREFIX)) {
+          break;
+        }
+        AppAttemptId appAttemptId = parseDbAppAttemptPathsKey(key);
+        if (appAttemptId == null) {
+          dbKeysToBeRemoved.add(e.getKey());
+          break;
+        }
+        AppPathsInfo appPathsInfo = parseDBAppAttemptPathsValue(e.getValue(), appAttemptId);
+        if (appPathsInfo == null) {
+          dbKeysToBeRemoved.add(e.getKey());
+          break;
+        }
+        appsShuffleInfo.compute(appAttemptId.appId,
+            (appId, existingAppShuffleInfo) -> {
+              if (existingAppShuffleInfo == null ||
+                  existingAppShuffleInfo.attemptId < appAttemptId.attemptId) {
+                dbKeysToBeRemoved.add(getDbAppAttemptPathsKey(
+                    new AppAttemptId(
+                        existingAppShuffleInfo.appId, existingAppShuffleInfo.attemptId)));
+                return new AppShuffleInfo(
+                    appAttemptId.appId, appAttemptId.attemptId, appPathsInfo);
+              } else {
+                dbKeysToBeRemoved.add(e.getKey());
+                return existingAppShuffleInfo;
+              }
+            });
+      }
+    }
+  }
+
+  private void reloadFinalizedAppAttemptsShuffleMergeInfo(DB db,  List<byte[]> dbKeysToBeRemoved) {
```

Comment:
same here. return keys to be removed instead of having the caller pass an empty collection